# Data Quality Audit of Raw Data from the ENTSO-E Transparency Platform

The main objective of the data quality audit is to assess the quality, consistency, and general characteristics of the raw data obtained from the ENTSO-E Transparency Platform.

The analysis focuses on the following aspects:
* Missing data and missing observations
* Index integrity and consistency
* Duplicated indices
* Timestamp and time zone consistency
* Overall data characteristics

**All data and variables fetched from ENTSO-E are explained in the [ENTSO-E data overview](./01_ENTSOE_data_overview.ipynb).**

In [ ]:
import sys
from pathlib import Path

import pandas as pd

sys.path.append(str(Path.cwd().parent.parent))
from src.config import (
    DATA_DIR,
    DOWNLOAD_START_DATE,
    DOWNLOAD_END_DATE,
    ENTSOE_BIDDING_ZONE_CODES,
)
from src.utils.data_quality import (
    print_index_summary,
    print_missing_values,
    print_data_overview,
    find_missing_dates,
    print_units_unavailability_analysis,
)

In [2]:
data_dir = DATA_DIR / "raw" / "entsoe"
date_range = f"{DOWNLOAD_START_DATE}_{DOWNLOAD_END_DATE}"
zone_codes = ENTSOE_BIDDING_ZONE_CODES

In [3]:
data_entose = {
    bz: {
        file.stem: pd.read_parquet(file)
        for file in (data_dir / bz / date_range).iterdir()
    }
    for bz in ENTSOE_BIDDING_ZONE_CODES
}

## 1. Poland (PL) data

### 1.1 Missing Data

In [4]:
data_poland = data_entose["PL"]
for name, data in data_poland.items():
    print_missing_values(data=data, name=name)


--- [load] Missing values ---
No missing values.

--- [generation] Missing values ---
Other renewable                              13529
Other                                        13529
Hydro Pumped Storage - Actual Consumption    32829
Wind Offshore                                85721

--- [load_forecast] Missing values ---
No missing values.

--- [day_ahead_prices] Missing values ---
No missing values.

--- [generation_wind_solar_forecast] Missing values ---
Wind Offshore    83993

--- [generation_units_unavailability] Missing values ---
docstatus        20045
nominal_power        1

--- [production_units_unavailability] Missing values ---
docstatus    847

--- [generation_forecast] Missing values ---
No missing values.


**Generation** 

In [5]:
missing_generation_types = [
    "Other renewable",
    "Other",
    "Hydro Pumped Storage - Actual Consumption",
    "Wind Offshore",
]

poland_generation = data_poland["generation"]

poland_missing_generation = poland_generation[missing_generation_types].copy()

poland_missing_generation["total_generation"] = poland_generation.sum(axis=1)

poland_missing_generation["missing_generation_total"] = poland_missing_generation[
    missing_generation_types
].sum(axis=1)

poland_missing_generation["missing_generation_share_pct"] = (
    poland_missing_generation["missing_generation_total"]
    / poland_missing_generation["total_generation"]
).round(2) * 100

print(
    f"Mean share of total generation from missing generation types: "
    f'{poland_missing_generation["missing_generation_share_pct"].mean().round(2)}%'
)
poland_missing_generation

Mean share of total generation from missing generation types: 3.74%


,Other renewable,Other,Hydro Pumped Storage - Actual Consumption,Wind Offshore,total_generation,missing_generation_total,missing_generation_share_pct
2022-12-01 00:00:00+01:00,NaN,NaN,NaN,NaN,18124.850,0.000,0.0
2022-12-01 01:00:00+01:00,NaN,NaN,NaN,NaN,17878.200,0.000,0.0
2022-12-01 02:00:00+01:00,NaN,NaN,NaN,NaN,17780.410,0.000,0.0
2022-12-01 03:00:00+01:00,NaN,NaN,NaN,NaN,17754.070,0.000,0.0
2022-12-01 04:00:00+01:00,NaN,NaN,NaN,NaN,18131.320,0.000,0.0
...,...,...,...,...,...,...,...
2026-09-15 22:45:00+02:00,54.998,552.439,6.737,87.6,17387.193,701.774,4.0
2026-09-15 23:00:00+02:00,54.492,547.993,8.339,110.4,17564.268,721.224,4.0
2026-09-15 23:15:00+02:00,54.155,544.366,8.339,134.8,17418.495,741.660,4.0
2026-09-15 23:30:00+02:00,54.048,544.348,12.090,156.8,17379.289,767.286,4.0


**Unavailability of generation/production units**

In [6]:
print(
    f"generation units:\n{data_poland["generation_units_unavailability"]["docstatus"].fillna("N/A").value_counts()}"
)
print(
    f"production units:\n{data_poland["production_units_unavailability"]["docstatus"].fillna("N/A").value_counts()}"
)

generation units:
docstatus
N/A    20045
Name: count, dtype: int64
production units:
docstatus
N/A    847
Name: count, dtype: int64


### 1.2 Missing Data – Observations

Missing values occur only in `generation`, `generation_wind_solar_forecast` and the two unavailability tables. Most gaps are structural (a series not reported yet), not random loss.

`Other`, `Other renewable` and `Hydro Pumped Storage - Actual Consumption` are NaN at the start of the sample and populated later; `Wind Offshore` has values only at the end (first offshore sources went live in July 2026). Combined, these four columns contribute on average only 3.74% of total generation (over the whole sample), confirming they can be dropped without a significant loss of signal.

In `generation_wind_solar_forecast`, only `Wind Offshore` is missing, and its forecast series starts earlier than the actual offshore output.

`docstatus` in both unavailability tables is 100% empty by construction (data pulled with `docStatus=A05` - Active), so it carries no information and can be dropped outright. The single missing `nominal_power` belongs to a record with no `plant_type`, location or name and must be excluded.

### 1.3 Index Integrity

In [7]:
for name, data in data_poland.items():
    print_index_summary(data=data, name=name)


--- [load] Index summary ---
Duplicates: 0
Duplicates due to Summer/Winter time switch: 9
Timezone:   Europe/Warsaw

Time steps:
  0 days 00:15:00: 79191
  0 days 01:00:00: 13441

--- [generation] Index summary ---
Duplicates: 0
Duplicates due to Summer/Winter time switch: 9
Timezone:   Europe/Warsaw

Time steps:
  0 days 00:15:00: 79191
  0 days 01:00:00: 13441

--- [load_forecast] Index summary ---
Duplicates: 0
Duplicates due to Summer/Winter time switch: 9
Timezone:   Europe/Warsaw

Time steps:
  0 days 00:15:00: 79191
  0 days 01:00:00: 13441

--- [day_ahead_prices] Index summary ---
Duplicates: 45
Duplicates due to Summer/Winter time switch: 6
Timezone:   Europe/Warsaw

Time steps:
  0 days 00:15:00: 33600
  0 days 01:00:00: 24839
  0 days 00:00:00: 45

--- [generation_wind_solar_forecast] Index summary ---
Duplicates: 0
Duplicates due to Summer/Winter time switch: 9
Timezone:   Europe/Warsaw

Time steps:
  0 days 00:15:00: 79191
  0 days 01:00:00: 13441

--- [generation_units_u

**Day-ahead prices**

In [8]:
poland_day_ahead_prices = data_poland["day_ahead_prices"].copy()
duplicated_poland_day_ahead_prices = poland_day_ahead_prices[
    poland_day_ahead_prices.index.duplicated(keep=False)
]
duplicated_poland_day_ahead_prices

,day_ahead_prices
2023-01-01 00:00:00+01:00,18.09
2023-01-01 00:00:00+01:00,18.09
2023-02-01 00:00:00+01:00,85.00
2023-02-01 00:00:00+01:00,85.00
2023-03-01 00:00:00+01:00,134.13
...,...
2026-07-01 00:00:00+02:00,151.12
2026-08-01 00:00:00+02:00,184.07
2026-08-01 00:00:00+02:00,184.07
2026-09-01 00:00:00+02:00,172.41


**Generation/Production units unavailability**

In [9]:
print("Poland - generation_units_unavailability")
print_units_unavailability_analysis(data=data_poland["generation_units_unavailability"])

Poland - generation_units_unavailability
Initial number of rows: 20045
Duplicated rows: 803
Duplicated (mrid, revision) rows: 0

Rows with created_doc_time more than 3 years before start:
businesstype
Planned maintenance    16
Name: count, dtype: int64

Rows with created_doc_time on or after end:
businesstype
Planned maintenance    8832
Unplanned outage       1758
Name: count, dtype: int64

Final number of rows: 8636

Planned maintenance:
Number of rows: 7962
created_doc_time range: 2022-02-03 17:18:05+00:00 - 2026-09-28 09:53:11+00:00

Unplanned outage:
Number of rows: 674
created_doc_time range: 2025-11-10 15:30:49+00:00 - 2026-09-17 21:24:21+00:00


,created_doc_time,avail_qty,biddingzone_domain,businesstype,curvetype,docstatus,end,mrid,nominal_power,plant_type,production_resource_id,production_resource_location,production_resource_name,production_resource_psr_name,pstn,qty_uom,resolution,revision,start
15,2022-02-03 18:18:05+01:00,0,PL,Planned maintenance,A03,None,2024-07-31 00:00:00+02:00,PX2OrSj18uuuUy_Pqnl6Ng,116.6,Fossil Hard coal,19W000000000076Y,Polska,Katowice,Katowice B1,1,MAW,PT1M,1,2024-07-01 00:01:00+02:00
17,2022-05-05 17:00:12+02:00,0,PL,Planned maintenance,A03,None,2024-11-16 00:00:00+01:00,m0uluObxl-bpf3eP2yVvmw,88.0,Fossil Hard coal,19W0000000000806,Polska,Kraków Łęg,Kraków Łęg B4,1,MAW,PT1M,2,2024-06-19 00:01:00+02:00
18,2025-10-06 13:38:03+02:00,0,PL,Planned maintenance,A03,None,2026-05-13 00:00:00+02:00,1AllRBDwFq5t8-WJ1I9WwQ,208.0,Biomass,19W000000000165Z,Polska,Połaniec 2-Pasywna,Połaniec 2 blok 9,1,MAW,PT1M,1,2026-04-17 00:01:00+02:00
1010,2025-10-07 01:28:02+02:00,0,PL,Planned maintenance,A03,None,2025-10-17 00:00:00+02:00,vQKoBwOpzGoNsbM4vRYkDA,88.9,Fossil Hard coal,19W0000000002191,Polska,EC Siekierki,Siekierki B09,1,MAW,PT1M,2,2025-07-28 00:01:00+02:00
1012,2025-10-07 01:28:14+02:00,0,PL,Planned maintenance,A03,None,2026-08-28 00:00:00+02:00,TCJod7mQ5DATJrP0yAUsOg,105.9,Fossil Hard coal,19W0000000002191,Polska,EC Siekierki,Siekierki B08,1,MAW,PT1M,2,2026-03-02 00:01:00+01:00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
19236,2026-09-24 08:10:28+02:00,0,PL,Planned maintenance,A03,None,2026-10-12 00:00:00+02:00,9Eyg1zvDemiERfsZwap6nA,221.0,Fossil Hard coal,19W0000000001519,Polska,Połaniec,Połaniec B7,1,MAW,PT1M,11,2026-08-29 19:01:00+02:00
19237,2026-09-24 10:31:20+02:00,0,PL,Planned maintenance,A03,None,2026-09-28 00:00:00+02:00,yfwejCTZis8jNUQgIK1BEA,834.0,Fossil Hard coal,19W000000000296K,Polska,Jaworzno 2 JWCD,Jaworzno 2 B7,1,MAW,PT1M,4,2026-07-06 00:04:00+02:00
19239,2026-09-24 14:42:37+02:00,0,PL,Planned maintenance,A03,None,2026-10-02 00:00:00+02:00,i4HyacDD3S6AerZd_0v0pQ,226.0,Fossil Brown coal/Lignite,19W000000000199I,Polska,Turów,Turów B03,1,MAW,PT1M,5,2026-09-12 00:00:00+02:00
19240,2026-09-26 13:12:03+02:00,0,PL,Planned maintenance,A03,None,2026-09-27 00:00:00+02:00,Mztc_GNS8kjyDkNcJOgfbw,213.0,Fossil Hard coal,19W000000000095U,Polska,Kozienice 1,Kozienice 1 B4,1,MAW,PT1M,6,2026-09-09 00:14:00+02:00


In [10]:
print("Poland - production_units_unavailability")
print_units_unavailability_analysis(data=data_poland["production_units_unavailability"])

Poland - production_units_unavailability
Initial number of rows: 847
Duplicated rows: 29
Duplicated (mrid, revision) rows: 0

Rows with created_doc_time more than 3 years before start:
businesstype
Planned maintenance    1
Name: count, dtype: int64

Rows with created_doc_time on or after end:
businesstype
Unplanned outage       469
Planned maintenance    341
Name: count, dtype: int64

Final number of rows: 7

Planned maintenance:
Number of rows: 7
created_doc_time range: 2020-11-30 12:08:35+00:00 - 2025-10-06 23:28:26+00:00

Unplanned outage:
Number of rows: 0
created_doc_time range: no data


,created_doc_time,avail_qty,biddingzone_domain,businesstype,curvetype,docstatus,end,mrid,nominal_power,plant_type,production_resource_id,production_resource_location,production_resource_name,production_resource_psr_name,pstn,qty_uom,resolution,revision,start
0,2020-11-30 13:08:35+01:00,441,PL,Planned maintenance,A03,None,2023-08-07 00:00:00+02:00,1EVbWTR-_UqYKfFmsx4JKA,620.0,Fossil Hard coal,19W0000000002191,Polska,EC Siekierki,,1,MAW,PT1M,1,2023-07-24 00:00:00+02:00
1,2020-11-30 13:08:40+01:00,473,PL,Planned maintenance,A03,None,2023-09-02 00:00:00+02:00,i1SD7GhEWqqBlFLqTo2yVw,620.0,Fossil Hard coal,19W0000000002191,Polska,EC Siekierki,,1,MAW,PT1M,1,2023-08-26 00:00:00+02:00
2,2020-11-30 13:11:24+01:00,441,PL,Planned maintenance,A03,None,2023-08-26 00:00:00+02:00,gtO_PlcCvVYzP5cqVfDUNQ,620.0,Fossil Hard coal,19W0000000002191,Polska,EC Siekierki,,1,MAW,PT1M,1,2023-08-07 00:00:00+02:00
3,2020-11-30 13:12:35+01:00,504,PL,Planned maintenance,A03,None,2023-09-16 00:00:00+02:00,-_xiswbu6tOJtq-NZKmj_g,620.0,Fossil Hard coal,19W0000000002191,Polska,EC Siekierki,,1,MAW,PT1M,1,2023-09-02 00:00:00+02:00
75,2025-10-07 01:28:19+02:00,505,PL,Planned maintenance,A03,None,2026-08-18 00:00:00+02:00,5az3Te-5EIACvlWo89amXg,533.8,Fossil Hard coal,19W0000000002191,Polska,EC Siekierki,,1,MAW,PT1M,1,2026-06-29 00:01:00+02:00
76,2025-10-07 01:28:19+02:00,473,PL,Planned maintenance,A03,None,2026-06-29 00:00:00+02:00,8fQR97ll9FqGiLr8G_QZxw,533.8,Fossil Hard coal,19W0000000002191,Polska,EC Siekierki,,1,MAW,PT1M,2,2026-06-08 00:01:00+02:00
77,2025-10-07 01:28:26+02:00,194,PL,Planned maintenance,A03,None,2026-07-14 00:00:00+02:00,xPpwLkHyVoCZ9mjgt206yw,245.8,Fossil Hard coal,19W000000000297I,Polska,EC Żerań 1,,1,MAW,PT1M,1,2026-07-01 00:01:00+02:00


### 1.4 Index Integrity – Observations

All PL series are tz-aware (Europe/Warsaw), no timestamps are missing. Resolution is mixed: `load`, `generation` and the forecast tables share one index (13,441 hourly + 79,191 15-min steps); `day_ahead_prices` stayed hourly much longer (24,839 hourly steps), so the 2025 MTU change affects only prices. A common grid is needed before merging.

DST repeats (9 timestamps, 6 in prices) are valid distinct instants, not duplicates. `day_ahead_prices` has 45 genuine exact duplicates, always at 00:00 on the first day of a month, and should be deduplicated.

Unavailability tables need real cleaning, not just deduplication on the index. `generation_units_unavailability` shows 7,149 duplicated index values, but only 803 of them are exact duplicate rows; the rest are different outage records sharing a `created_doc_time`, so index-level dedup alone is wrong - full-row dedup is required. No `(mrid, revision)` pairs repeat outside the exact duplicates, confirming revisions are used correctly.

The dominant data-loss driver is not duplication but the way of publishing: 10,590 of 20,045 PL generation-unit records (and 810 of 847 production-unit records) have `created_doc_time` on or after the outage's own `end`, meaning the report was created only after the event was already over. A further 16 records have `created_doc_time` implausibly more than 3 years before `start` and should be dropped as bad metadata.

After removing duplicates, PL keeps 8,636 usable generation-unit records (7,962 planned maintenance, 674 unplanned outage) but only 7 usable production-unit records, all planned maintenance. For generation units every surviving "unplanned outage" record was created no earlier than 2025-11-10, so this category has no usable history before that date, while "planned maintenance" covers the sample from 2022 onwards. This asymmetry is the basis for keeping only planned-maintenance records as a feature.

### 1.5 Data Descriptive Statistics

In [11]:
for name, data in data_poland.items():
    print_data_overview(data=data, name=name)


--- [load] Overview ---
Shape: 92633 x 1

Info:
<class 'pandas.DataFrame'>
DatetimeIndex: 92633 entries, 2022-12-01 00:00:00+01:00 to 2026-09-15 23:45:00+02:00
Data columns (total 1 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Actual Load  92633 non-null  float64
dtypes: float64(1)
memory usage: 1.4 MB

Description:
        Actual Load
count  92633.000000
mean   18305.505318
std     3152.694048
min    10463.962000
25%    15753.941000
50%    18411.141000
75%    20451.363000
max    28303.880000

--- [generation] Overview ---
Shape: 92633 x 15

Info:
<class 'pandas.DataFrame'>
DatetimeIndex: 92633 entries, 2022-12-01 00:00:00+01:00 to 2026-09-15 23:45:00+02:00
Data columns (total 15 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   Biomass                                    92633 non-null  float64
 1   Fossil Brown coal/Lignite

            Biomass  Fossil Brown coal/Lignite  Fossil Coal-derived gas  \
count  92633.000000               92633.000000             92633.000000   
mean     303.949663                3589.136399               149.397911   
std       86.161133                1072.705949                43.108341   
min        0.000000                 902.873000                 9.000000   
25%      255.740000                2763.174000               136.574000   
50%      313.740000                3637.664000               160.791000   
75%      369.897000                4435.401000               180.221000   
max      454.637000                6557.615000               226.894000   

         Fossil Gas  Fossil Hard coal    Fossil Oil  Hydro Pumped Storage  \
count  92633.000000      92633.000000  92633.000000          92633.000000   
mean    2182.963557       6289.817030    185.140519            130.509217   
std      809.232546       1836.949925     62.921216            243.509467   
min      247.230

## 2. Other European Bidding Zones included data

In [12]:
data_entose = {bz: dataset for bz, dataset in data_entose.items() if bz != "PL"}

### 2.1 Missing Data

In [13]:
for bz, datasets_bz in data_entose.items():
    print(f"\n--- {bz} " + "-" * 40)
    for name, data in datasets_bz.items():
        print_missing_values(data=data, name=name)


--- AT ----------------------------------------

--- [load_forecast] Missing values ---
No missing values.

--- [water_reservoirs_hydro_storage] Missing values ---
No missing values.

--- [day_ahead_prices] Missing values ---
No missing values.

--- [generation_wind_solar_forecast] Missing values ---
No missing values.

--- [generation_forecast] Missing values ---
No missing values.

--- CZ ----------------------------------------

--- [load_forecast] Missing values ---
No missing values.

--- [day_ahead_prices] Missing values ---
No missing values.

--- [generation_wind_solar_forecast] Missing values ---
No missing values.

--- [generation_forecast] Missing values ---
No missing values.

--- DE_LU ----------------------------------------

--- [load_forecast] Missing values ---
No missing values.

--- [day_ahead_prices] Missing values ---
No missing values.

--- [generation_wind_solar_forecast] Missing values ---
Solar           12
Wind Onshore    12

--- [generation_units_unavailabil

**DE_LU - generation wind and solar forecast**

In [14]:
print("[DE_LU] - missing values for generation_wind_solar_forecast\n")

missing_wind_onshore_de_lu = data_entose["DE_LU"]["generation_wind_solar_forecast"][
    "Wind Onshore"
][data_entose["DE_LU"]["generation_wind_solar_forecast"]["Wind Onshore"].isna()]
print("\nWind Onshore")
print(f"total missing: {len(missing_wind_onshore_de_lu)}\n")
print(missing_wind_onshore_de_lu.index)

missing_solar_de_lu = data_entose["DE_LU"]["generation_wind_solar_forecast"]["Solar"][
    data_entose["DE_LU"]["generation_wind_solar_forecast"]["Solar"].isna()
]
print("\nSolar")
print(f"total missing: {len(missing_solar_de_lu)}\n")
print(missing_solar_de_lu.index)

[DE_LU] - missing values for generation_wind_solar_forecast


Wind Onshore
total missing: 12

DatetimeIndex(['2023-10-29 00:00:00+02:00', '2023-10-29 00:15:00+02:00',
               '2023-10-29 00:30:00+02:00', '2023-10-29 00:45:00+02:00',
               '2024-10-27 00:00:00+02:00', '2024-10-27 00:15:00+02:00',
               '2024-10-27 00:30:00+02:00', '2024-10-27 00:45:00+02:00',
               '2025-10-26 00:00:00+02:00', '2025-10-26 00:15:00+02:00',
               '2025-10-26 00:30:00+02:00', '2025-10-26 00:45:00+02:00'],
              dtype='datetime64[us, Europe/Berlin]', freq=None)

Solar
total missing: 12

DatetimeIndex(['2023-10-29 00:00:00+02:00', '2023-10-29 00:15:00+02:00',
               '2023-10-29 00:30:00+02:00', '2023-10-29 00:45:00+02:00',
               '2024-10-27 00:00:00+02:00', '2024-10-27 00:15:00+02:00',
               '2024-10-27 00:30:00+02:00', '2024-10-27 00:45:00+02:00',
               '2025-10-26 00:00:00+02:00', '2025-10-26 00:15:00+02:00',
            

**FR - generation wind and solar forecast**

In [15]:
print("[FR] - missing values for generation_wind_solar_forecast\n")

missing_wind_onshore_fr = data_entose["FR"]["generation_wind_solar_forecast"][
    "Wind Onshore"
][data_entose["FR"]["generation_wind_solar_forecast"]["Wind Onshore"].isna()]
print("\nWind Onshore")
print(f"total missing: {len(missing_wind_onshore_fr)}\n")
print(missing_wind_onshore_fr.index)

missing_wind_offshore_fr = data_entose["FR"]["generation_wind_solar_forecast"][
    "Wind Offshore"
][data_entose["FR"]["generation_wind_solar_forecast"]["Wind Offshore"].isna()]
print("\nWind Offshore")
print(f"total missing: {len(missing_wind_offshore_fr)}\n")
print(missing_wind_offshore_fr.index)

missing_solar_fr = data_entose["FR"]["generation_wind_solar_forecast"]["Solar"][
    data_entose["FR"]["generation_wind_solar_forecast"]["Solar"].isna()
]
print("\nSolar")
print(f"total missing: {len(missing_solar_fr)}\n")
print(missing_solar_fr.index)

[FR] - missing values for generation_wind_solar_forecast


Wind Onshore
total missing: 27

DatetimeIndex(['2023-10-29 23:00:00+01:00', '2024-10-27 23:00:00+01:00',
               '2025-10-26 23:00:00+01:00', '2026-01-02 00:00:00+01:00',
               '2026-01-02 01:00:00+01:00', '2026-01-02 02:00:00+01:00',
               '2026-01-02 03:00:00+01:00', '2026-01-02 04:00:00+01:00',
               '2026-01-02 05:00:00+01:00', '2026-01-02 06:00:00+01:00',
               '2026-01-02 07:00:00+01:00', '2026-01-02 08:00:00+01:00',
               '2026-01-02 09:00:00+01:00', '2026-01-02 10:00:00+01:00',
               '2026-01-02 11:00:00+01:00', '2026-01-02 12:00:00+01:00',
               '2026-01-02 13:00:00+01:00', '2026-01-02 14:00:00+01:00',
               '2026-01-02 15:00:00+01:00', '2026-01-02 16:00:00+01:00',
               '2026-01-02 17:00:00+01:00', '2026-01-02 18:00:00+01:00',
               '2026-01-02 19:00:00+01:00', '2026-01-02 20:00:00+01:00',
               '2026-01-02 21:00:

**HU - generation wind and solar forecast**

In [16]:
print("[HU] - missing values generation_wind_solar_forecast")
missing_wind_onshore_hu = data_entose["HU"]["generation_wind_solar_forecast"][
    "Wind Onshore"
][data_entose["HU"]["generation_wind_solar_forecast"]["Wind Onshore"].isna()]
print("\nWind Onshore")
print(f"total missing: {len(missing_wind_onshore_hu)}\n")
print(missing_wind_onshore_hu.index)

[HU] - missing values generation_wind_solar_forecast

Wind Onshore
total missing: 34

DatetimeIndex(['2026-02-16 15:30:00+01:00', '2026-02-16 15:45:00+01:00',
               '2026-02-16 16:00:00+01:00', '2026-02-16 16:15:00+01:00',
               '2026-02-16 16:30:00+01:00', '2026-02-16 16:45:00+01:00',
               '2026-02-16 17:00:00+01:00', '2026-02-16 17:15:00+01:00',
               '2026-02-16 17:30:00+01:00', '2026-02-16 17:45:00+01:00',
               '2026-02-16 18:00:00+01:00', '2026-02-16 18:15:00+01:00',
               '2026-02-16 18:30:00+01:00', '2026-02-16 18:45:00+01:00',
               '2026-02-16 19:00:00+01:00', '2026-02-16 19:15:00+01:00',
               '2026-02-16 19:30:00+01:00', '2026-02-16 19:45:00+01:00',
               '2026-02-16 20:00:00+01:00', '2026-02-16 20:15:00+01:00',
               '2026-02-16 20:30:00+01:00', '2026-02-16 20:45:00+01:00',
               '2026-02-16 21:00:00+01:00', '2026-02-16 21:15:00+01:00',
               '2026-02-16 21:30:00+01

**LT - generation forecast**

In [17]:
print("[LT] - generation forecast data schema")
data_entose["LT"]["generation_forecast"]

[LT] - generation forecast data schema


,Scheduled Consumption,Scheduled Aggregated,Actual Aggregated
2022-12-01 01:00:00+02:00,670.0,278.87,NaN
2022-12-01 02:00:00+02:00,889.0,277.37,NaN
2022-12-01 03:00:00+02:00,886.0,276.47,NaN
2022-12-01 04:00:00+02:00,882.0,276.67,NaN
2022-12-01 05:00:00+02:00,670.0,367.87,NaN
...,...,...,...
2026-09-15 23:45:00+03:00,0.0,290.90,NaN
2026-09-16 00:00:00+03:00,0.0,286.90,NaN
2026-09-16 00:15:00+03:00,0.0,298.50,NaN
2026-09-16 00:30:00+03:00,0.0,310.00,NaN


### 2.2 Missing Data – Observations

DE_LU and FR wind/solar forecast gaps line up with the three autumn DST-change Sundays (DE_LU: first hour; FR: last hour), so they are DST artefacts, not random loss. FR `Wind Offshore` is missing as one block from the sample start to 2023-08-06 — structural, forecasts simply start later. FR `Solar` and `Wind Onshore` are also missing for the whole of 2026-01-02. HU has one isolated 8.5-hour gap in `Wind Onshore` on 2026-02-16.

LT `generation_forecast` uses a different schema (three columns instead of one) and `Actual Aggregated` is mostly empty - not usable in its current form.

DE_LU and FR unavailability tables show the same empty `docstatus` pattern as PL. 77 DE_LU production-unit records have no `nominal_power` and must be dropped.

### 2.3 Index Integrity

In [18]:
for bz, datasets_bz in data_entose.items():
    print(f"\n--- {bz} " + "-" * 40)
    for name, data in datasets_bz.items():
        print_index_summary(data=data, name=name)


--- AT ----------------------------------------

--- [load_forecast] Index summary ---
Duplicates: 0
Duplicates due to Summer/Winter time switch: 12
Timezone:   Europe/Vienna

Time steps:
  0 days 00:15:00: 132955

--- [water_reservoirs_hydro_storage] Index summary ---
Duplicates: 0
Duplicates due to Summer/Winter time switch: 0
Timezone:   UTC

Time steps:
  7 days 00:00:00: 194
  14 days 00:00:00: 1

--- [day_ahead_prices] Index summary ---
Duplicates: 0
Duplicates due to Summer/Winter time switch: 6
Timezone:   Europe/Vienna

Time steps:
  0 days 00:15:00: 33598
  0 days 01:00:00: 24835
  0 days 02:00:00: 2
  0 days 00:30:00: 1

--- [generation_wind_solar_forecast] Index summary ---
Duplicates: 0
Duplicates due to Summer/Winter time switch: 12
Timezone:   Europe/Vienna

Time steps:
  0 days 00:15:00: 132955

--- [generation_forecast] Index summary ---
Duplicates: 0
Duplicates due to Summer/Winter time switch: 3
Timezone:   Europe/Vienna

Time steps:
  0 days 00:15:00: 28795
  0 day

**DE_LU - generation/production units unavailability**

In [19]:
print_units_unavailability_analysis(
    data=data_entose["DE_LU"]["generation_units_unavailability"]
)

Initial number of rows: 62028
Duplicated rows: 20879
Duplicated (mrid, revision) rows: 21852

Rows with created_doc_time more than 3 years before start:
businesstype
Planned maintenance    1
Name: count, dtype: int64

Rows with created_doc_time on or after end:
businesstype
Planned maintenance    11238
Unplanned outage        5123
Name: count, dtype: int64

Final number of rows: 2935

Planned maintenance:
Number of rows: 2305
created_doc_time range: 2019-11-15 12:41:08+00:00 - 2026-09-29 15:14:15+00:00

Unplanned outage:
Number of rows: 630
created_doc_time range: 2021-07-08 09:24:47+00:00 - 2026-09-18 16:05:50+00:00


,created_doc_time,avail_qty,biddingzone_domain,businesstype,curvetype,docstatus,end,mrid,nominal_power,plant_type,production_resource_id,production_resource_location,production_resource_name,production_resource_psr_name,pstn,qty_uom,resolution,revision,start
0,2019-11-15 13:41:08+01:00,0,DE_LU,Planned maintenance,A03,None,2022-01-03 06:00:00+01:00,E8YyC47mHZcL5gr9HlTX4g,405.0,Fossil Gas,11WD7GERS-X-KW-C,intra_zonal,Gersteinwerk,Gersteinwerk I,1,MAW,PT1M,1,2022-01-01 00:00:00+01:00
1,2021-01-06 14:21:32+01:00,0,DE_LU,Planned maintenance,A03,None,2023-08-07 05:00:00+02:00,3h-AmzgTAkp-1O3yGflUBQ,870.0,Fossil Gas,11WD7EMSL-G-KW-T,intra_zonal,Emsland,Emsland D,1,MAW,PT1M,1,2023-07-28 23:00:00+02:00
2,2021-04-06 12:48:06+02:00,116,DE_LU,Planned maintenance,A03,None,2023-02-05 21:30:00+01:00,TVqLZrx2_fKjNuinbZ_NyA,475.0,Fossil Gas,11WD7EMSL-G-KW-T,intra_zonal,Emsland,Emsland B,1,MAW,PT1M,1,2023-02-03 21:30:00+01:00
3,2021-04-06 12:49:32+02:00,116,DE_LU,Planned maintenance,A03,None,2023-02-12 21:30:00+01:00,uMvoLm1YWYtVQ5-fTmUGKQ,475.0,Fossil Gas,11WD7EMSL-G-KW-T,intra_zonal,Emsland,Emsland B,1,MAW,PT1M,1,2023-02-10 21:30:00+01:00
4,2021-04-06 12:49:33+02:00,116,DE_LU,Planned maintenance,A03,None,2023-05-07 21:30:00+02:00,3jjpEG4-OQHYES-nfp9msg,475.0,Fossil Gas,11WD7EMSL-G-KW-T,intra_zonal,Emsland,Emsland B,1,MAW,PT1M,1,2023-05-05 21:30:00+02:00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
19289,2026-09-29 10:24:04+02:00,0,DE_LU,Planned maintenance,A03,None,2026-10-01 00:00:00+02:00,0n-inN6olePdfEgCzGOdwA,608.0,Fossil Gas,11W0-0000-0708-1,intra_zonal,Herne,GUD_HERNE_6,1,MAW,PT1M,4,2026-09-12 01:00:00+02:00
19292,2026-09-29 13:59:31+02:00,0,DE_LU,Planned maintenance,A03,None,2027-01-01 00:00:00+01:00,4ogKMEyyq-3Jwoi3z5hjBw,211.0,Fossil Hard coal,11WD7VOLK5S--KWI,North Rhine-Westphalia,Voelklingen,VOELKLINGEN_HKV,1,MAW,PT1M,3,2026-06-01 00:00:00+02:00
19294,2026-09-29 14:49:17+02:00,0,DE_LU,Planned maintenance,A03,None,2026-10-19 12:00:00+02:00,9XkmSZecVN7cG0vo-j3ybQ,333.0,Fossil Hard coal,11WD2NO2K000308Z,intra_zonal,Nord 2,Nord 2 T20,1,MAW,PT1M,3,2026-04-06 11:00:00+02:00
19295,2026-09-29 15:01:13+02:00,5,DE_LU,Planned maintenance,A03,None,2026-10-03 00:00:00+02:00,6TexJL4fBdgnIRCBt_A_jA,90.0,Fossil Gas,11W0-0000-0696-D,Hesse,KMW_GENERATION,KMW_KW2GT,1,MAW,PT1M,3,2026-08-29 00:00:00+02:00


In [20]:
print_units_unavailability_analysis(
    data=data_entose["DE_LU"]["production_units_unavailability"]
)

Initial number of rows: 17773
Duplicated rows: 2958
Duplicated (mrid, revision) rows: 3193

Rows with created_doc_time more than 3 years before start:
Series([], Name: count, dtype: int64)

Rows with created_doc_time on or after end:
businesstype
Planned maintenance    9861
Unplanned outage       1111
Name: count, dtype: int64

Final number of rows: 650

Planned maintenance:
Number of rows: 495
created_doc_time range: 2025-10-07 12:48:56+00:00 - 2026-09-16 18:35:45+00:00

Unplanned outage:
Number of rows: 155
created_doc_time range: 2025-11-11 14:06:14+00:00 - 2026-09-12 21:50:39+00:00


,created_doc_time,avail_qty,biddingzone_domain,businesstype,curvetype,docstatus,end,mrid,nominal_power,plant_type,production_resource_id,production_resource_location,production_resource_name,production_resource_psr_name,pstn,qty_uom,resolution,revision,start
4639,2025-10-07 14:48:56+02:00,110,DE_LU,Planned maintenance,A03,None,2025-12-10 12:00:00+01:00,wsSOuejSqA7UhWzZEzlvHA,220.0,Hydro Pumped Storage,11WD2ERZH0002682,intra_zonal,PSW Erzhausen,,1,MAW,PT1M,1,2025-12-10 10:00:00+01:00
5637,2025-10-08 01:50:52+02:00,2018,DE_LU,Planned maintenance,A03,None,2026-01-01 00:00:00+01:00,yXNzPBbaK-9n3-qO_CJmIA,2142.0,Hydro Pumped Storage,11WD43VIWXHOILLM,intra_zonal,Vorarlberger Illwerke AG,,197821,MAW,PT1M,10,2025-12-12 16:00:00+01:00
8204,2025-11-10 16:27:41+01:00,335,DE_LU,Planned maintenance,A03,None,2025-11-11 01:00:00+01:00,GIjKUrz6UZ2xBN0pHe8RsA,400.0,Wind Offshore,11WD2OGTI000264J,Nordsee AWZ,Global Tech I,,1,MAW,PT1M,1,2025-11-10 20:15:00+01:00
8246,2025-11-11 15:06:14+01:00,281,DE_LU,Unplanned outage,A03,None,2025-11-11 16:00:00+01:00,5fIPZtJkYMaXbCvNLS_4dA,396.0,Wind Offshore,11W0-0000-0095-9,intra_zonal,Merkur Offshore GmbH,,1,MAW,PT1M,1,2025-11-11 15:00:00+01:00
8260,2025-11-12 00:40:38+01:00,320,DE_LU,Planned maintenance,A03,None,2025-11-12 01:00:00+01:00,jOy8A7M1-Tbrg_Z7gXUBGQ,400.0,Wind Offshore,11WD2OGTI000264J,Nordsee AWZ,Global Tech I,,1,MAW,PT1M,1,2025-11-12 00:30:00+01:00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
11614,2026-09-12 00:25:24+02:00,0,DE_LU,Planned maintenance,A03,None,2026-09-15 02:00:00+02:00,TSz7iw0qY-uz7huTPKyXxA,400.0,Wind Offshore,11WD2OGTI000264J,Nordsee AWZ,Global Tech I,,1,MAW,PT1M,1,2026-09-14 02:00:00+02:00
11615,2026-09-12 23:50:39+02:00,31,DE_LU,Unplanned outage,A03,None,2026-09-13 00:00:00+02:00,zBeTzyARDSJoldUm7rpjqA,200.0,Wind Offshore,11W0-0000-0261-N,intra_zonal,Trianel Windpark Borkum II,,1,MAW,PT1M,1,2026-09-12 23:30:00+02:00
11618,2026-09-14 12:39:30+02:00,0,DE_LU,Planned maintenance,A03,None,2026-09-15 12:00:00+02:00,4TJN2jv8Ohb9EpFQNdC3OA,252.0,Wind Offshore,11WD2OGW20003378,intra_zonal,Gode Wind II,,1,MAW,PT1M,1,2026-09-15 07:00:00+02:00
11620,2026-09-16 11:52:46+02:00,143,DE_LU,Planned maintenance,A03,None,2026-09-17 08:00:00+02:00,27rWHBL2VJLHEIUNzxzG3g,315.0,Wind Offshore,11W0-0000-0952-Q,intra_zonal,Windanker,,3031,MAW,PT1M,2,2026-09-16 19:00:00+02:00


**FR - generation/production units unavailability**

In [21]:
print_units_unavailability_analysis(
    data=data_entose["FR"]["generation_units_unavailability"]
)

Initial number of rows: 22997
Duplicated rows: 2961
Duplicated (mrid, revision) rows: 436

Rows with created_doc_time more than 3 years before start:
Series([], Name: count, dtype: int64)

Rows with created_doc_time on or after end:
businesstype
Planned maintenance    10865
Unplanned outage        5196
Name: count, dtype: int64

Final number of rows: 3539

Planned maintenance:
Number of rows: 2891
created_doc_time range: 2020-11-16 15:33:32+00:00 - 2026-09-29 13:08:00+00:00

Unplanned outage:
Number of rows: 648
created_doc_time range: 2021-04-26 06:41:34+00:00 - 2026-09-24 20:38:42+00:00


,created_doc_time,avail_qty,biddingzone_domain,businesstype,curvetype,docstatus,end,mrid,nominal_power,plant_type,production_resource_id,production_resource_location,production_resource_name,production_resource_psr_name,pstn,qty_uom,resolution,revision,start
0,2020-11-16 16:33:32+01:00,0,FR,Planned maintenance,A03,None,2023-09-29 17:00:00+02:00,J181q5_XX8Qc43_fZCxPSA,152.0,Hydro Pumped Storage,17W100P100P02756,FRANCE,GRAND MAISON,GRAND MAISON 7,1,MAW,PT1M,2,2023-08-28 07:00:00+02:00
1,2021-04-26 08:41:34+02:00,0,FR,Unplanned outage,A03,None,2025-01-01 00:00:00+01:00,rJz-2tM-YqQLxt3Gyj9RNQ,595.0,Fossil Hard coal,17W100P100P03469,France,PROVENCE 5,PROVENCE 5,1,MAW,PT1M,1,2024-01-01 00:00:00+01:00
2,2021-04-26 08:41:35+02:00,0,FR,Unplanned outage,A03,None,2024-01-01 00:00:00+01:00,mVtCmqx_qT3_PbZO2PBwCg,595.0,Fossil Hard coal,17W100P100P03469,France,PROVENCE 5,PROVENCE 5,1,MAW,PT1M,1,2023-01-01 00:00:00+01:00
3,2021-04-26 08:41:35+02:00,0,FR,Unplanned outage,A03,None,2023-01-01 00:00:00+01:00,SBBDlApI938XfEA0GwgaNQ,595.0,Fossil Hard coal,17W100P100P03469,France,PROVENCE 5,PROVENCE 5,1,MAW,PT1M,1,2022-01-01 00:00:00+01:00
4,2021-05-11 17:35:25+02:00,0,FR,Planned maintenance,A03,None,2023-07-21 17:00:00+02:00,Uq4Bwu38eL_QHjrxIwqR0A,170.0,Hydro Pumped Storage,17W100P100P02756,FRANCE,GRAND MAISON,GRAND MAISON 10,1,MAW,PT1M,1,2023-07-17 07:00:00+02:00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
19595,2026-09-28 09:49:08+02:00,0,FR,Planned maintenance,A03,None,2026-10-03 23:00:00+02:00,gO1p-Zebazg7Dj9l3UJRhw,1300.0,Nuclear,17W100P100P0190E,FRANCE,CATTENOM 4,CATTENOM 4,1,MAW,PT1M,25,2026-01-31 00:00:00+01:00
19596,2026-09-28 15:10:41+02:00,0,FR,Planned maintenance,A03,None,2026-10-13 00:00:00+02:00,P1vNS7aGnQhZroGSt_0jcg,1500.0,Nuclear,17W100P100P01954,FRANCE,CHOOZ 1,CHOOZ 1,1,MAW,PT1M,25,2026-08-01 21:30:00+02:00
19597,2026-09-28 15:11:19+02:00,0,FR,Planned maintenance,A03,None,2026-10-13 00:00:00+02:00,N5Jj5ZcayWZ3kLX1XG4COg,1500.0,Nuclear,17W100P100P01962,FRANCE,CHOOZ 2,CHOOZ 2,1,MAW,PT1M,38,2026-07-10 23:45:00+02:00
19598,2026-09-29 12:02:49+02:00,0,FR,Planned maintenance,A03,None,2026-10-08 17:00:00+02:00,bIMMKzHPUKpyCawXbyVhuA,182.0,Hydro Water Reservoir,17W100P100P0300X,FRANCE,VILLARODIN,VILLARODIN 1,1,MAW,PT1M,7,2026-04-24 17:00:00+02:00


In [22]:
print_units_unavailability_analysis(
    data=data_entose["FR"]["production_units_unavailability"]
)

Initial number of rows: 4626
Duplicated rows: 307
Duplicated (mrid, revision) rows: 0

Rows with created_doc_time more than 3 years before start:
Series([], Name: count, dtype: int64)

Rows with created_doc_time on or after end:
businesstype
Planned maintenance    2763
Unplanned outage        589
Name: count, dtype: int64

Final number of rows: 967

Planned maintenance:
Number of rows: 917
created_doc_time range: 2021-03-01 07:15:25+00:00 - 2026-09-28 14:42:38+00:00

Unplanned outage:
Number of rows: 50
created_doc_time range: 2025-11-16 20:18:18+00:00 - 2026-09-15 20:38:36+00:00


,created_doc_time,avail_qty,biddingzone_domain,businesstype,curvetype,docstatus,end,mrid,nominal_power,plant_type,production_resource_id,production_resource_location,production_resource_name,production_resource_psr_name,pstn,qty_uom,resolution,revision,start
0,2021-03-01 08:15:25+01:00,0,FR,Planned maintenance,A03,None,2022-12-06 12:00:00+01:00,qu4tWf-urs6ir3Pp_h3KBQ,116.0,Hydro Water Reservoir,17W100P100P0299T,FRANCE,SAINT GUILLERME,,1,MAW,PT1M,1,2022-12-06 09:00:00+01:00
1,2021-09-14 15:50:22+02:00,0,FR,Planned maintenance,A03,None,2023-03-24 17:00:00+01:00,sJtZwc9cWd3JK8yJe5og3A,150.0,Hydro Run-of-river and pondage,17W100P100P0309F,FRANCE,SAUSSAZ,,1,MAW,PT1M,2,2023-02-27 07:00:00+01:00
2,2022-02-07 18:05:22+01:00,0,FR,Planned maintenance,A03,None,2023-01-17 16:00:00+01:00,2hMCvWRD8-6c0fbeh8cvEg,297.0,Hydro Water Reservoir,17W100P100P02853,FRANCE,MALGOVERT,,1,MAW,PT1M,1,2023-01-17 07:00:00+01:00
3,2022-02-11 14:35:26+01:00,0,FR,Planned maintenance,A03,None,2023-06-06 12:00:00+02:00,c4xMtpxO5bXqpYaj85pDnw,116.0,Hydro Water Reservoir,17W100P100P0299T,FRANCE,SAINT GUILLERME,,1,MAW,PT1M,1,2023-06-06 09:00:00+02:00
4,2022-02-21 15:05:27+01:00,0,FR,Planned maintenance,A03,None,2022-12-30 23:59:00+01:00,rsBgKxN4BRqlv2e6Bin0XA,420.0,Fossil Gas,17W100P100P0318E,France,FR-GA-MORANT1,,1,MAW,PT1M,1,2022-12-24 00:00:00+01:00
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4310,2026-09-15 14:03:22+02:00,0,FR,Unplanned outage,A03,None,2026-09-15 22:00:00+02:00,gUZ1TrSrS2_nOnIPsDbQyQ,435.0,Fossil Gas,17W100P100P0345B,FRANCE,SAINT AVOLD 8,,1,MAW,PT1M,2,2026-09-15 06:18:00+02:00
4311,2026-09-15 22:38:36+02:00,0,FR,Unplanned outage,A03,None,2026-09-16 07:00:00+02:00,A7fj2QVixamtzsxH0KZg6w,435.0,Fossil Gas,17W100P100P0345B,FRANCE,SAINT AVOLD 8,,1,MAW,PT1M,1,2026-09-15 22:37:00+02:00
4314,2026-09-24 14:39:45+02:00,218,FR,Planned maintenance,A03,None,2026-09-25 17:00:00+02:00,hGpxKhx4DrBiB1NhbYOrHw,384.0,Hydro Water Reservoir,17W100P100P02926,FRANCE,SERRE PONCON,,1,MAW,PT1M,2,2026-08-28 17:00:00+02:00
4315,2026-09-24 17:34:04+02:00,0,FR,Planned maintenance,A03,None,2026-10-20 23:59:00+02:00,krOoHsJMfu9uUojDnlkwVg,435.0,Fossil Gas,17W100P100P0344D,FRANCE,SAINT AVOLD 7,,1,MAW,PT1M,1,2026-07-15 00:00:00+02:00


In [24]:
for bz, datasets_bz in data_entose.items():
    print(f"\n--- {bz} " + "-" * 40)
    for name, data in datasets_bz.items():
        find_missing_dates(data=data, name=name, freq="h")


--- AT ----------------------------------------

--- [load_forecast] Missing timestamps (freq=h) ---
Count: 0

--- [day_ahead_prices] Missing timestamps (freq=h) ---
Count: 3

Missing ranges:
  2023-11-30 00:00:00+01:00 (1 timestamp)
  2024-11-30 00:00:00+01:00 (1 timestamp)
  2025-11-30 00:00:00+01:00 (1 timestamp)

Missing timestamps by day of week:
  Thursday: 1
  Saturday: 1
  Sunday: 1

--- [generation_wind_solar_forecast] Missing timestamps (freq=h) ---
Count: 0

--- [generation_forecast] Missing timestamps (freq=h) ---
Count: 0

--- CZ ----------------------------------------

--- [load_forecast] Missing timestamps (freq=h) ---
Count: 0

--- [day_ahead_prices] Missing timestamps (freq=h) ---
Count: 3

Missing ranges:
  2023-11-30 00:00:00+01:00 (1 timestamp)
  2024-11-30 00:00:00+01:00 (1 timestamp)
  2025-11-30 00:00:00+01:00 (1 timestamp)

Missing timestamps by day of week:
  Thursday: 1
  Saturday: 1
  Sunday: 1

--- [generation_wind_solar_forecast] Missing timestamps (freq=

### 2.4 Index Integrity – Observations

Resolution and switch timing differ by zone and dataset; a common UTC grid is required before any cross-zone merge. `day_ahead_prices` is the one exception with an identical structure across all non-PL zones.

The 00:00 timestamp is missing on 30 November (2023–2025) in every non-PL zone at the same UTC instant, pointing to a systematic cause (download or publication window) rather than zone-specific outages — a single, predictable gap to interpolate. LT `generation_forecast` is missing 290 hourly timestamps in 5 gaps, the longest lasting 8 days (193 hours), and stays excluded. DE_LU `load_forecast` and `generation_forecast` are missing the first hour of the autumn DST-change Sundays, as in the wind/solar forecast. Whole-day gaps remain in FR (`generation_wind_solar_forecast`, `generation_forecast`), SK (`generation_forecast`) and SE_4 (`load_forecast`, `generation_wind_solar_forecast`, `generation_forecast`; the longest lasts 2 days). FR `generation_forecast`, LT `load_forecast` and LT `generation_wind_solar_forecast` additionally contain isolated gaps of up to 1 hour.

Unavailability filtering confirms the PL pattern but at larger scale. DE_LU generation-unit records: 62,028 initial rows, 20,879 exact duplicates, and 21,852 `(mrid, revision)` duplicates. After removing duplicates, implausible timestamps and retrospective records, 2,935 rows remain (2,305 planned, 630 unplanned). FR keeps 3,539 of 22,997 generation-unit rows (2,891 planned, 648 unplanned).

Production-unit tables shrink even harder: DE_LU keeps 650 of 17,773 rows, FR keeps 967 of 4,626 - confirming this table adds limited value versus `generation_units_unavailability` once properly filtered.

## 3. Conclusions and Data Cleaning / Preprocessing Decisions

**Main findings**
PL core series have no missing timestamps; remaining NaN values are structural or DST-related. Unavailability tables need three cleaning layers: exact-duplicate removal, `(mrid, revision)` deduplication, and removal of retrospective records (`created_doc_time >= end`) — this last filter causes most of the row loss in every zone. In PL, unplanned-outage records exist only since November 2025, whereas in DE_LU and FR they are available earlier; planned maintenance covers a longer history everywhere. LT `generation_forecast` is unusable.

### 3.1 Data Cleaning Decisions

**Poland (PL)**

* Drop `Other`, `Other renewable`, `Hydro Pumped Storage - Actual Consumption` and `Wind Offshore` from `generation`; drop `Wind Offshore` from `generation_wind_solar_forecast`.
* Deduplicate `day_ahead_prices`, keeping the first record.
* Unavailability tables: drop `docstatus`; cast `avail_qty` and `pstn` to numeric; drop records with missing `nominal_power` or blank `plant_type`; keep only `businesstype = Planned maintenance`; do not use `production_units_unavailability` due to the low number of records; drop exact duplicate rows and `(mrid, revision)` duplicates; drop records where `created_doc_time` is more than 3 years before `start`; drop records where `created_doc_time >= end`.

**Germany (DE_LU)**

* Use `generation_units_unavailability` with both `Planned maintenance` and `Unplanned outage`.
* Do not use `production_units_unavailability`, as its data starts only from 2025-10.
* Apply the same unavailability cleaning procedure as for PL, deduplicating before aggregation.

**France (FR)**

* Drop `Wind Offshore` from `generation_wind_solar_forecast`.
* Use `generation_units_unavailability` with both `Planned maintenance` and `Unplanned outage`.
* From `production_units_unavailability`, keep only `Planned maintenance`.
* Apply the same unavailability cleaning procedure and deduplicate before aggregation.

**Lithuania (LT)**

* Exclude `generation_forecast` entirely, as the series is not usable.

### 3.2 Data Preprocessing Decisions

**Time Index**

* Convert all relevant time indices to UTC before merging, comparing or aggregating data across bidding zones.
* Perform `day_ahead_prices` deduplication after conversion to UTC.

**Poland (PL)**

* No interpolation is needed for the PL core series — no missing timestamps. Remaining NaN values are treated as structural/DST-related, not as gaps to impute.

**Other bidding zones**

* Gaps of up to 3 hours – linear interpolation on the UTC grid: the missing 30 November 00:00 price timestamp in all non-PL zones, DST-related gaps in DE_LU/FR and isolated gaps in FR `generation_forecast`, LT `load_forecast` and LT `generation_wind_solar_forecast`.
* Longer gaps (more than 3 hours) in weather-driven series (`generation_wind_solar_forecast`) – copy the same hours from the previous available day (D-1): FR `Solar`/`Wind Onshore` (2026-01-02), HU `Wind Onshore` (2026-02-16, 15:30–23:45), the whole-day gap in FR (2023-04-18) and the five whole-day gaps in SE_4 (for the 2-day gap use the last available day before it).
* Whole-day gaps in demand- and schedule-driven series – copy the same hours from one week earlier (D-7): SE_4 `load_forecast` (2025-12-04), `generation_forecast` in FR (2025-03-06), SK (2023-12-08) and SE_4 (2025-12-02).
* D-1 is used for weather-driven series because weather persists from one day to the next, and D-7 for demand- and schedule-driven series because they follow a weekly (weekday/weekend) cycle. The hours are matched by local time of day. The 3-hour limit keeps interpolation within a small part of the daily cycle (solar, load); the gaps found are either up to 1 hour or at least 8.5 hours long, so no case lies close to the limit.
* Reservoir data (AT, NO, SE): forward fill only, including the 14-day step.
